# Project 3: Baby Mood Prediction via Deep Neural Networks

## Scientific Question: Can prediction algorithms be utilized to predict baby moods, such as, happiness, sadness, fright, disgust, hunger, thirst, and sleepiness from neural frequencies?

Babies cry a lot and for many different reasons, sometimes we know what they want, but other times we are left in the dust. It would be nice to know how they feel and since we cannot talk with them, it would be even better if we could look inside their brains and just know what they want! That is where my prediction algorithm and I come in. We know that whenever you, a baby, or I want something there is neural activity taking place. Electrodes and EEGs are devices and test that allow us to examine electrical charges and activity in specific regions of the brain. With the use of electrodes this neural activity can be monitored and can grasp this neural activity in the form of high gamma and low frequencies. These frequencies correspond with regions of the brain and respectively, with different feelings, emotions, and important to us, moods.

For this experiment we are utilizing the MNIST database, this database consists of 70,000 black and white images of reconstructed baby brain complexes via machine learning algorithms that utilized the neural activation and respective frequency for specific brain region and complex reconstruction in the form of 28 x 28 pixelated images. The data on the frequencies utilized for reconstruction came from implanting electrodes directly onto the babies' brains and retrieving information based on their known moods. The data came out to 60,000 images used for training and 10,000 images used for testing.

# Scientific Hypothesis: If babies' moods and their respective neural activation and frequency are correlated with specific parts of the brain, then we can utilize a prediction algorithm to examine reconstructed neural activity in the form of brain images to predict brain complexes and their respective mood association. 

Analysis was performed by clustering the data into known and unknown, given by their respective training and testing denotation. This data was fed into a Deep Neural Network prediction algorithm and propagated through to examine, measure, and increase prediction accuracy. We observe respective loss and prediction accuracy through each epoch (or pass through). We are distinctly examining the Convolution Neural Network's error, to illuminate error reduction desires we compare and contrast each epoch and different levels of complexity for the Deep Neural Network (CNN), both for a simple and complex model, respectively. 

To answer this scientific question as well as test my hypothesis I needed to locate a dataset that had collected neural frequencies from babies based on their moods and reconstructed brain parts and complexes based on those collected frequencies. Fortunately, this data (proof of concept) is part of the Keras library and can be imported directly from there once adding the Keras library to your environment.

## Part 1: Load the Packages: 

- Keras is an API or an application programming interface, this particular interface is used specifically in Python and creates a means for utilizing artificial neural networks. Keras handles weight calculation and simplifies layer input. It notably carries the dataset used here, MNIST.
- Tensorflow is a library used for machine learning, it has tools, datasets, libraries, packages, and resources that help with deployment of Machine Learning. Keras has Tensorflow wrapped inside of it and is how Keras is able to function and compute on the back end. Notably, it is not directly imported because Keras runs over it.

In [ ]:
# Here we are importing all the packages and subpackages that are necessary to run our Deep Neural Network
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Dropout
from keras.layers import Flatten
from keras.layers.convolutional import Conv2D
from keras.layers.convolutional import MaxPooling2D
from keras.utils import np_utils

## Part 2: Load the Data, split it into Testing and Training, and Reshape it: 

The MNIST data is loaded in, since the dataset is split into testing and training sets, we first need to classify the clusters in accordance to how they appear in the dataframe. The data needs to be reshaped via linear transformations by the samples, width, height, and channels. The data needs to be converted to 32 bit in order to allow it to be processed by our ram more easily. Similarly the data must be normalized, this ids done by dividing our test and training data by 225 in accordance to the max byte size value, doing this will help keep our data between 0 and 1, which is necessary for the inputs and outputs and for computing weights as they are passed through and propagated by the Network. We utilize our utilities in order to take the integers and make them binary, this is done with one hot and is specifically used for our y outputs. This is done as at the end of our network layer, the total layer is formed of 10 neurons that all add to 1 in value, the value with the highest number is the network's "prediction neuron" as I like to describe it.

In [15]:
# Data is loaded in with respect to its cluster as a part of the training or testing set, the x are the inputs and
#the y values are the outputs
(X_train, y_train), (X_test, y_test) = mnist.load_data()
# The data needs to be reshaped in the form of the samples, width, height, and channels. 
X_train = X_train.reshape((X_train.shape[0], 28, 28, 1)).astype('float32')
X_test = X_test.reshape((X_test.shape[0], 28, 28, 1)).astype('float32')
# The inputs are normalized so that they are integers between 0 and 1, this is doen by dividing by 255.
X_train = X_train / 255
X_test = X_test / 255
# The outputs, the y values for train and test, the integers are made binary.
y_train = np_utils.to_categorical(y_train)
y_test = np_utils.to_categorical(y_test)
num_classes = y_test.shape[1]

## Part 3: Define and Create the Simple Convolution Neural Network Model:

Now, we build the architecture of our Neural Network. We define and create the model, whereby the sequential format is best for our particular model data. The sequential model is useful for our architecture as it allows us to build each layer individually, this is particularly useful for us as we are keeping this neural network concise and efficient. We are making our network a Convolution Neural Network as this form of neural network works well for analyzing images and is especially useful for the 28x28 image constructions passed through. Since we are using a sequential model, we can choose our activation desired for each layer. Relu refers to activations that pass weighted values only if they are positive, negative values will be fixed and passed as 0 (so no negative weights are passed), this aids in learning and prediction efficiency and performance. The softmax activation is used in our final layer to illustrate the network’s probability distribution and pushes the results to either 0 or 1. Other layers will have different activations and uses as will be annotated in the code below.

In [23]:
# Construction of the Simple Convolution Neural Network model and architecture
def baseline_model():
	# The model's architecture is build and layers, from input to output are added
    # The sequential model is our model type as it allows for us to add each layer individually
	model = Sequential()
    # First Convolution layer is added and is with respect to our inputs, we are using 32 convolution filters, and 5 kernel rows
    # and columns. One input is 1 28x28 pixelated image. The activation relu is used here as described above.
	model.add(Conv2D(32, (5, 5), input_shape=(28, 28, 1), activation='relu'))
    # This layer is used to filter and reduced the parameters as they are passed through the CNN.
	model.add(MaxPooling2D())
    # Dropouts are added to prevent overfitting, a value of 1 means there will be no dropout and closer to 0 means no output.
    # Here I elected to used 0.2 because this is our input layer and we must have more leniency towards the beginning.
	model.add(Dropout(0.2))
    # The flatten layer flattens the weights from previous convolution layers before the layers become dense,
    # this makes the values 1 dimensional and are therefor more easily passed to the first fully connected dense layer.
	model.add(Flatten())
    # A dense layer of 128 neurons is added which also uses the relu activation
	model.add(Dense(128, activation='relu'))
    # Our second dense layer is added with neurons correlated to the number of classes in our y output test group.
    # Softmax activation is notably used here (as described above).
	model.add(Dense(num_classes, activation='softmax'))
	# The model is compiled and we classify our loss, our optimizer (adamn, which is useful for stochastic gradients),
    # and our metric which is accuracy and is used to examine performance.
    # The categorical crossentropy is a useful loss alongside softmax activation, the loss function is ideal for one hot vectors.
	model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
	return model

## Part 4: Build, Fit, and Evaluate the model: 

In this section of the code we are fitting the model and evaluating its performance for our analysis. In this section we will run the code and illustrate each of the 10 epochs for this particular network, it will show us the time it took for the epoch to run, the loss associated with that epoch, the accuracy as well as the loss and accuracy for the validation data set. For our data analysis we will also be examining the score of the DNN which is indicated by the Error as well as showcasing the loss and accuracy [loss,accuracy]; we will print a summary of our model indicating each layer, its output shape, and respective parameters.  

In [25]:
# Reestablish our model from that which was defined above.
model = baseline_model()
# Fit the model, showcasing that our data is fit by the training sets input x and output y (training) and the validation set
# the testing set, validation test for x and y. Epochs = number of passes through the DNN/CNN for training, batch size = sample size
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=10, batch_size=200)
# Evaluation of the Simple CNN, print score in the form of CNN error
scores = model.evaluate(X_test, y_test, verbose=0)
print(scores)
print("CNN Error: %.2f%%" % (100-scores[1]*100))
# Summarizes the model by layer, output shape, and parameters.
model.summary()

Epoch 1/10
300/300 [==============================] - 8s 26ms/step - loss: 0.4792 - accuracy: 0.8605 - val_loss: 0.0797 - val_accuracy: 0.9755
Epoch 2/10
300/300 [==============================] - 8s 26ms/step - loss: 0.0796 - accuracy: 0.9772 - val_loss: 0.0515 - val_accuracy: 0.9827
Epoch 3/10
300/300 [==============================] - 8s 26ms/step - loss: 0.0532 - accuracy: 0.9842 - val_loss: 0.0478 - val_accuracy: 0.9847
Epoch 4/10
300/300 [==============================] - 8s 26ms/step - loss: 0.0389 - accuracy: 0.9880 - val_loss: 0.0368 - val_accuracy: 0.9881
Epoch 5/10
300/300 [==============================] - 8s 26ms/step - loss: 0.0297 - accuracy: 0.9912 - val_loss: 0.0317 - val_accuracy: 0.9893
Epoch 6/10
300/300 [==============================] - 8s 27ms/step - loss: 0.0261 - accuracy: 0.9915 - val_loss: 0.0307 - val_accuracy: 0.9896
Epoch 7/10
300/300 [==============================] - 8s 26ms/step - loss: 0.0198 - accuracy: 0.9942 - val_loss: 0.0325 - val_accuracy: 0.9888

## Part 5: Compare this Simple Neural Network to a Deep, more Complex One: 

Here I am showcasing a larger Convolution Neural Network that is deeper than the first simple layer, the deeper a network, the more layers that the network has. This is to showcase how we can improve the prediction capabilities of the network by adding more layers and increasing complexity to reduce loss, increase accuracy, and overall decrease percent error for the network. Notably, we added another Convolution 2D layer with a relu activation as well as a second dense layer but this time only 50 neurons. I will not be describing each line here, because the network is very similar, but I will annotate what layers were added.

In [30]:
# Construction of the Larger Convolution Neural Network model and architecture
def larger_model():
	# Building the Architecture
	model2 = Sequential()
	model2.add(Conv2D(30, (5, 5), input_shape=(28, 28, 1), activation='relu'))
	model2.add(MaxPooling2D())
    # Second Convolution layer is addedusing 15 convolution filters, and 3 kernel rows and columns, also using relu activation.
	model2.add(Conv2D(15, (3, 3), activation='relu'))
    # Another filtering layer is added for reduced parameterization.
	model2.add(MaxPooling2D())
	model2.add(Dropout(0.2))
	model2.add(Flatten())
	model2.add(Dense(128, activation='relu'))
    # Another dense layer is added this time with 50 neurons, still using a relu activation.
	model2.add(Dense(50, activation='relu'))
	model2.add(Dense(num_classes, activation='softmax'))
	# Model is compiled with the same loss, optimization, and metric for evaluation.
	model2.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
	return model2

In [31]:
# Construct the model
model = larger_model()
# Fit the model and run the epochs
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=10, batch_size=200)

Epoch 1/10
300/300 [==============================] - 9s 30ms/step - loss: 0.8186 - accuracy: 0.7424 - val_loss: 0.0837 - val_accuracy: 0.9736
Epoch 2/10
300/300 [==============================] - 9s 30ms/step - loss: 0.1068 - accuracy: 0.9680 - val_loss: 0.0570 - val_accuracy: 0.9813
Epoch 3/10
300/300 [==============================] - 9s 31ms/step - loss: 0.0819 - accuracy: 0.9746 - val_loss: 0.0416 - val_accuracy: 0.9862
Epoch 4/10
300/300 [==============================] - 9s 29ms/step - loss: 0.0615 - accuracy: 0.9810 - val_loss: 0.0356 - val_accuracy: 0.9883
Epoch 5/10
300/300 [==============================] - 9s 30ms/step - loss: 0.0500 - accuracy: 0.9846 - val_loss: 0.0348 - val_accuracy: 0.9882
Epoch 6/10
300/300 [==============================] - 9s 30ms/step - loss: 0.0445 - accuracy: 0.9863 - val_loss: 0.0315 - val_accuracy: 0.9886
Epoch 7/10
300/300 [==============================] - 9s 29ms/step - loss: 0.0385 - accuracy: 0.9883 - val_loss: 0.0346 - val_accuracy: 0.9883

In [33]:
# Model Evaluation and New CNN error score
scores = model.evaluate(X_test, y_test, verbose=0)
print(scores)
print("Large CNN Error: %.2f%%" % (100-scores[1]*100))
model.summary()

[0.02396288700401783, 0.9926000237464905]
Large CNN Error: 0.74%
Model: "sequential_11"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_14 (Conv2D)           (None, 24, 24, 30)        780       
_________________________________________________________________
max_pooling2d_14 (MaxPooling (None, 12, 12, 30)        0         
_________________________________________________________________
conv2d_15 (Conv2D)           (None, 10, 10, 15)        4065      
_________________________________________________________________
max_pooling2d_15 (MaxPooling (None, 5, 5, 15)          0         
_________________________________________________________________
dropout_11 (Dropout)         (None, 5, 5, 15)          0         
_________________________________________________________________
flatten_11 (Flatten)         (None, 375)               0         
______________________________________________________

## Part 6: Analysis of Results

The substance of our results come from the prediction accuracy and error that we observe through each of the epochs of the Neural Networks as well as their overall values. The results I see here prove my scientific hypothesis. A Deep Neural Network is able to analyze reconstructed images of babies' brains made from frequencies related to their specific mood and is able to accurately predict what part of the brain was reconstructed and can correlate that prediction to a specific mood. We can see that as the epochs pass, for both networks, the accuracy increases, and the loss decreases quite significantly. The simple neural network showcases an accuracy of 99.04% and a loss of 2.8%, with an error of 0.95%. So we can safely say that a strong majority of the time our model can look at the reconstructed images of the brains and determine what specific complex it is and determine what the mood is based on association. Furthermore, evaluating the larger Convolution Neural Network allows to see notable improvements, showcasing an accuracy of 99.3% a loss of 2.4%, with an overall error of just 0.74%. This indicates that by deepening our network and increasing its complexity we can improve our models prediction, though this seems slight, the closer we are able to get to perfect, the better; I would rather know more confidently what the babies mood is so that we can live in peace and understand with them.

Sources and help: machinelearning.com, elitedatascience.com , DataCamp, and Professor Schiffer